pm/sender.go

```diff
@@ -0,0 +1,82 @@
+package pm
+
+import (
+	"math/big"
+	"sync"
+	"sync/atomic"
+
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/livepeer/go-livepeer/eth"
+	"github.com/pkg/errors"
+)
+
+// Sender enables starting multiple probabilistic micropayment sessions with multiple recipients
+// and create tickets that adhere to each session's params and unique nonce requirements.
+type Sender interface {
+	StartSession(recipient ethcommon.Address, ticketParams TicketParams) string
+
+	CreateTicket(sessionID string) (*Ticket, *big.Int, []byte, error)
+
+	// Later: support receiving new recipientRandHash values mid-stream
+}
+
+type session struct {
+	senderNonce uint64
+
+	recipient ethcommon.Address
+
+	ticketParams TicketParams
+}
+
+type defaultSender struct {
+	accountManager eth.AccountManager
+
+	sessions sync.Map
+}
+
+// NewSender creates a new Sender instance.
+func NewSender(accountManager eth.AccountManager) Sender {
+	return &defaultSender{
+		accountManager: accountManager,
+	}
+}
+
+func (s *defaultSender) StartSession(recipient ethcommon.Address, ticketParams TicketParams) string {
+	sessionID := hashToHex(ticketParams.RecipientRandHash)
+
+	s.sessions.Store(sessionID, &session{
+		recipient:    recipient,
+		ticketParams: ticketParams,
+		senderNonce:  0,
+	})
+
+	return sessionID
+}
+
+func (s *defaultSender) CreateTicket(sessionID string) (*Ticket, *big.Int, []byte, error) {
+	recipientRandHash := hexToHash(sessionID)
+
+	tempSession, ok := s.sessions.Load(sessionID)
+	if !ok {
+		return nil, nil, nil, errors.Errorf("cannot create a ticket for an unknown session: %+v", sessionID)
```

Comment:
That could work. However, before I go for it, I'm curious as to how you would resolve the string const with the idea of adding runtime variable values to the error string?

For example, if the const string were `"cannot create a ticket for an unknown session: %+v"`, it would not really help us with testing.

An alternative I can think of is to concatenate the const error message and the variable in question, but I was having resistance to this approach because it feels unidiomatic.

wdyt?

Reply:
Now that we've discussed the idea of erring on the side of detailed errors for now - would love to get your take on if and how you would change this line of code?

Reply:
I think if we're using dynamic error strings, we'll have to use string matching while testing anyway so I think the way this error is returned seems fine now.

However, I don't think we need to use `%+v` here - IIUC that verb is only necessary when we need to log an error along with its stack trace of wrapped errors according to https://dave.cheney.net/2016/06/12/stack-traces-and-the-errors-package

Reply:
yeah, I think you're right. getting rid of the `%+v`.